File: csharp/src/DataStructures/HashTable.cs
namespace DSA.DataStructures;

/// <summary>
/// Hash table implementation with open addressing and linear probing.
/// Time: O(1) average for all operations
/// Space: O(n)
/// </summary>
public class HashTable<TKey, TValue> where TKey : notnull
{
    private class Entry
    {
        public TKey Key { get; }
        public TValue Value { get; set; }
        public bool Deleted { get; set; }

        public Entry(TKey key, TValue value)
        {
            Key = key;
            Value = value;
        }
    }

    private Entry?[] _buckets;
    private int _size;
    private const double LoadFactorThreshold = 0.7;

    public int Count => _size;
    public bool IsEmpty => _size == 0;

    public HashTable(int initialCapacity = 16)
    {
        _buckets = new Entry?[Math.Max(1, initialCapacity)];
    }

    public void Put(TKey key, TValue value)
    {
        if (_size >= _buckets.Length * LoadFactorThreshold)
            Resize(_buckets.Length * 2);

        var index = GetIndex(key);
        var firstDeleted = -1;

        for (int i = 0; i < _buckets.Length; i++)
        {
            var entry = _buckets[index];

            if (entry == null)
            {
                var insertIndex = firstDeleted != -1 ? firstDeleted : index;
                _buckets[insertIndex] = new Entry(key, value);
                _size++;
                return;
            }

            if (entry.Deleted && firstDeleted == -1)
            {
                firstDeleted = index;
            }
            else if (!entry.Deleted && EqualityComparer<TKey>.Default.Equals(entry.Key, key))
            {
                entry.Value = value;
                return;
            }

            index = (index + 1) % _buckets.Length;
        }

        if (firstDeleted != -1)
        {
            _buckets[firstDeleted] = new Entry(key, value);
            _size++;
        }
    }

    public TValue? Get(TKey key)
    {
        var index = FindIndex(key);
        return index == -1 ? default : _buckets[index]!.Value;
    }

    public bool TryGet(TKey key, out TValue? value)
    {
        var index = FindIndex(key);
        if (index == -1)
        {
            value = default;
            return false;
        }
        value = _buckets[index]!.Value;
        return true;
    }

    public TValue? Remove(TKey key)
    {
        var index = FindIndex(key);
        if (index == -1)
            return default;

        var entry = _buckets[index]!;
        var value = entry.Value;
        entry.Deleted = true;
        _size--;
        return value;
    }

    public bool ContainsKey(TKey key) => FindIndex(key) != -1;

    public List<TKey> Keys()
    {
        var result = new List<TKey>();
        foreach (var entry in _buckets)
        {
            if (entry != null && !entry.Deleted)
                result.Add(entry.Key);
        }
        return result;
    }

    public List<TValue> Values()
    {
        var result = new List<TValue>();
        foreach (var entry in _buckets)
        {
            if (entry != null && !entry.Deleted)
                result.Add(entry.Value);
        }
        return result;
    }

    public void Clear()
    {
        Array.Clear(_buckets);
        _size = 0;
    }

    private int GetIndex(TKey key) => Math.Abs(key.GetHashCode()) % _buckets.Length;

    private int FindIndex(TKey key)
    {
        var index = GetIndex(key);

        for (int i = 0; i < _buckets.Length; i++)
        {
            var entry = _buckets[index];

            if (entry == null)
                return -1;

            if (!entry.Deleted && EqualityComparer<TKey>.Default.Equals(entry.Key, key))
                return index;

            index = (index + 1) % _buckets.Length;
        }

        return -1;
    }

    private void Resize(int newCapacity)
    {
        var oldBuckets = _buckets;
        _buckets = new Entry?[newCapacity];
        _size = 0;

        foreach (var entry in oldBuckets)
        {
            if (entry != null && !entry.Deleted)
                Put(entry.Key, entry.Value);
        }
    }
}

File: php/src/DataStructures/HashTable.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Hash table implementation with open addressing and linear probing.
 * Time: O(1) average for all operations
 * Space: O(n)
 *
 * @template K
 * @template V
 */
class HashTable
{
    private array $buckets;
    private int $size = 0;
    private int $capacity;
    private const LOAD_FACTOR_THRESHOLD = 0.7;

    public function __construct(int $initialCapacity = 16)
    {
        $this->capacity = max(1, $initialCapacity);
        $this->buckets = array_fill(0, $this->capacity, null);
    }

    public function put(mixed $key, mixed $value): void
    {
        if ($this->size >= $this->capacity * self::LOAD_FACTOR_THRESHOLD) {
            $this->resize($this->capacity * 2);
        }

        $index = $this->hash($key);
        $firstDeleted = -1;

        for ($i = 0; $i < $this->capacity; $i++) {
            $entry = $this->buckets[$index];

            if ($entry === null) {
                $insertIndex = $firstDeleted !== -1 ? $firstDeleted : $index;
                $this->buckets[$insertIndex] = new HashEntry($key, $value);
                $this->size++;
                return;
            }

            if ($entry->deleted && $firstDeleted === -1) {
                $firstDeleted = $index;
            } elseif (!$entry->deleted && $entry->key === $key) {
                $entry->value = $value;
                return;
            }

            $index = ($index + 1) % $this->capacity;
        }

        if ($firstDeleted !== -1) {
            $this->buckets[$firstDeleted] = new HashEntry($key, $value);
            $this->size++;
        }
    }

    public function get(mixed $key): mixed
    {
        $index = $this->findIndex($key);
        return $index === -1 ? null : $this->buckets[$index]->value;
    }

    public function remove(mixed $key): mixed
    {
        $index = $this->findIndex($key);
        if ($index === -1) {
            return null;
        }
        $entry = $this->buckets[$index];
        $value = $entry->value;
        $entry->deleted = true;
        $this->size--;
        return $value;
    }

    public function contains(mixed $key): bool
    {
        return $this->findIndex($key) !== -1;
    }

    public function isEmpty(): bool
    {
        return $this->size === 0;
    }

    public function size(): int
    {
        return $this->size;
    }

    public function keys(): array
    {
        $result = [];
        foreach ($this->buckets as $entry) {
            if ($entry !== null && !$entry->deleted) {
                $result[] = $entry->key;
            }
        }
        return $result;
    }

    public function values(): array
    {
        $result = [];
        foreach ($this->buckets as $entry) {
            if ($entry !== null && !$entry->deleted) {
                $result[] = $entry->value;
            }
        }
        return $result;
    }

    public function clear(): void
    {
        $this->buckets = array_fill(0, $this->capacity, null);
        $this->size = 0;
    }

    private function hash(mixed $key): int
    {
        return abs(crc32((string)$key)) % $this->capacity;
    }

    private function findIndex(mixed $key): int
    {
        $index = $this->hash($key);

        for ($i = 0; $i < $this->capacity; $i++) {
            $entry = $this->buckets[$index];

            if ($entry === null) {
                return -1;
            }

            if (!$entry->deleted && $entry->key === $key) {
                return $index;
            }

            $index = ($index + 1) % $this->capacity;
        }

        return -1;
    }

    private function resize(int $newCapacity): void
    {
        $oldBuckets = $this->buckets;
        $this->capacity = $newCapacity;
        $this->buckets = array_fill(0, $this->capacity, null);
        $this->size = 0;

        foreach ($oldBuckets as $entry) {
            if ($entry !== null && !$entry->deleted) {
                $this->put($entry->key, $entry->value);
            }
        }
    }
}

/**
 * @internal
 */
class HashEntry
{
    public function __construct(
        public mixed $key,
        public mixed $value,
        public bool $deleted = false
    ) {}
}
